src/store/index-level.ts

```diff
@@ -1,213 +1,435 @@
+    // cursor indexes must match the provided filters in order to be valid.
+    // ie: if someone passes a valid messageCid for a cursor that's not part of the filter.
+    if (FilterUtility.matchAnyFilter(indexes, filters)) {
+      return IndexLevel.keySegmentJoin(IndexLevel.encodeValue(sortValue), itemId);
+    }
+  }
 
-    // delete all indexes associated with the data of the given ID
-    const ops: LevelWrapperBatchOperation<string>[] = [ ];
-    for (const propertyName in indexes) {
-      const propertyValue = indexes[propertyName];
-      const key = this.join(propertyName, this.encodeValue(propertyValue), dataId);
-      ops.push({ type: 'del', key });
+  /**
+   * Queries the provided searchFilters asynchronously, returning results that match the matchFilters.
+   *
+   * @param matchFilters the filters passed to the parent query.
+   * @param searchFilters the modified filters used for the LevelDB query to search for a subset of items to match against.
+   */
+  async queryWithInMemoryPaging(
+    tenant: string,
+    matchFilters: Filter[],
+    searchFilters:Filter[],
+    queryOptions: QueryOptions,
+    options?: IndexLevelOptions
+  ): Promise<string[]> {
+    const { sortProperty, sortDirection = SortDirection.Ascending, cursor, limit } = queryOptions;
+
+    // we create a matches map so that we can short-circuit matched items within the async single query below.
+    const matches:Map<string, IndexedItem> = new Map();
+
+    try {
+      await Promise.all(searchFilters.map(filter => {
+        return this.executeSingleFilterQuery(tenant, filter, matchFilters, sortProperty, matches, options );
+      }));
+    } catch (error) {
+      if ((error as DwnError).code === DwnErrorCode.IndexInvalidSortProperty) {
+        // return empty results if the sort property is invalid.
+        return [];
+      }
     }
 
-    ops.push({ type: 'del', key: `__${dataId}__indexes` });
+    const sortedValues = [...matches.values()].sort((a,b) => this.sortItems(a,b, sortProperty, sortDirection));
 
-    await partition.batch(ops, options);
+    // we find the cursor point and only return the result starting there + the limit.
+    // if there is no cursor index, we just start in the beginning.
+    const cursorIndex = cursor ? sortedValues.findIndex(match => match.itemId === cursor) : -1;
+    if (cursor !== undefined && cursorIndex === -1) {
+      // if a cursor is provided but we cannot find it, we return an empty result set
+      return [];
+    }
+
+    const start = cursorIndex > -1 ? cursorIndex + 1 : 0;
+    const end = limit !== undefined ? start + limit : undefined;
+
+    return sortedValues.slice(start, end).map(match => match.itemId);
   }
 
-  async clear(): Promise<void> {
-    return this.db.clear();
+  /**
+   * Execute a filtered query against a single filter and return all results.
+   */
+  private async executeSingleFilterQuery(
+    tenant: string,
+    searchFilter: Filter,
+    matchFilters: Filter[],
```

Comment:
Design: love the direction that you are going with a mix of db query and in-memory matches. Though it's odd here that the entire OR filters (`matchFilters`) "bleed" into every individual execution of each `searchFilter`.

Does it make sense to treat individual filters in the OR filters truly independently when executing them? We can still have the `searchFilter` but just singular `matchFilter` and have the filter properties being mutually exclusive for intent clarity.

This is probably related to the inner workings of `FilterSelector` and I have followed your instruction to not look at it at all, but maybe it's finally the right time to bring it into discussions?

Reply:
Good call out here, I was thinking there are a few different directions that we can go with optimizing this.

The one-to-one searchFilter/matchFilter is interesting, but I think it makes the `cursor` starting point a bit more complicated. The supplied `cursor` could have come exclusively from one of the `searchFilter->matchFilter` pair.

I was also thinking we could go the opposite way by attempting to bring down an array of `Filter` into a singular `Filter`. Reducing down any common fields into the existing `OneOfFilter`, the bigger change would be creating something like a `OneOfRange` to handle OR ranges (not in use practically in code yet).

So there are definitely a number of enhancements that could happen here, as well as `FilterSelector` itself.

Happy to discuss `FilterSelector` with you, basically I'd like to come up with a good starting place for some **rules** that it follows in order to take multiple filters, and reduce them into a common set of searchFilters, and in certain scenarios fall back to returning an empty filter array, which would go to db-query.

These rules can change over time as we find better heuristics... I was kind of playing around with some ideas but don't love on where I landed.

Reply:
Bounced ideas offline, will review again in the next iteration.